File: backend/app/schemas/eligibility.py
from typing import Any, Optional

from pydantic import BaseModel, Field


class ValidationRuleSchema(BaseModel):
    must_be_active: Optional[bool] = None
    must_be_unrestricted: Optional[bool] = None
    must_be_accredited: Optional[bool] = None
    license_type: Optional[str] = None
    certification_type: Optional[str] = None
    registration_type: Optional[str] = None
    degree_types: Optional[list[str]] = None
    allowed_levels: Optional[list[str]] = None
    required_for_specialties: Optional[list[str]] = None


class RequirementBase(BaseModel):
    requirement_type: str = Field(default="")
    name: str = Field(default="")
    description: str = Field(default="")
    is_required: bool = True
    validation_rules: dict[str, Any] = Field(default_factory=dict)
    base_requirement_id: Optional[int] = None
    provider_type_id: Optional[int] = None
    id: Optional[int] = None

    class Config:
        from_attributes = True


class RequirementCreate(RequirementBase):
    pass


class Requirement(BaseModel):
    id: int
    requirement_type: str
    name: str
    description: str
    is_required: bool
    base_requirement_id: int
    provider_type_id: int
    validation_rules: Optional[dict[str, Any]] = None

    class Config:
        from_attributes = True
        populate_by_name = True


class ProviderTypeBase(BaseModel):
    code: str
    name: str


class ProviderTypeCreate(ProviderTypeBase):
    requirements: list[RequirementCreate]


class ProviderType(BaseModel):
    id: int
    code: str
    name: str
    requirements: list[RequirementBase]

    class Config:
        from_attributes = True


class EligibilityCheck(BaseModel):
    npi: str


class EligibilityRequirements(BaseModel):
    stateLicense: bool
    deaCds: bool
    boardCertification: bool
    degree: bool
    residency: bool
    malpracticeInsurance: bool
    backgroundCheck: bool
    workHistory: bool
    providerType: str


class NPIValidation(BaseModel):
    providerName: str
    npi: str
    updateDate: str


class RawApiResponseInner(BaseModel):
    NPI_Validation: NPIValidation
    Licenses: list[dict[str, Any]]


class RawApiResponse(BaseModel):
    rawApiResponse: RawApiResponseInner


class EligibilityResponse(BaseModel):
    isEligible: bool
    requirements: dict[str, Any]
    rawApiResponse: dict[str, Any]

File: frontend/src/types/eligibility.ts
export interface ValidationRule {
  must_be_active?: boolean;
  must_be_unrestricted?: boolean;
  must_be_completed?: boolean;
  must_be_verified?: boolean;
  must_be_enrolled?: boolean;
  must_be_up_to_date?: boolean;
  license_type?: string;
  registration_type?: string;
  certification_type?: string;
  degree_type?: string;
  verification_period_years?: number;
  [key: string]: any;
}

export interface BaseRequirement {
  id: number;
  requirement_type: string;
  name: string;
  description: string;
  validation_rules: ValidationRule;
}

export interface ValidationDetail {
  issuer?: string;
  type?: string;
  number?: string;
  status?: string;
  expirationDate?: string | null;
  boardActions?: string[];
  hasBoardAction?: boolean;
  additionalInfo?: {
    deaSchedules?: string;
    licenseState?: string;
  };
  details?: {
    issuer?: string;
    type?: string;
    number?: string;
    status?: string;
    expirationDate?: string | null;
    boardActions?: string[];
    hasBoardAction?: boolean;
    additionalInfo?: {
      deaSchedules?: string;
      licenseState?: string;
    };
  };
}

export interface Requirement {
  id: number;
  requirement_type: string;
  type: string;
  name: string;
  description: string;
  validation_rules: ValidationRule;
  is_required: boolean;
  is_valid: boolean;
  validation_message?: string;
  status: 'valid' | 'invalid' | 'required' | 'optional';
  details: ValidationDetail[];
  base_requirement_id: number;
  provider_type_id: number;
  severity: number;
}

export interface ProviderType {
  id: number;
  code: string;
  name: string;
  requirements: Requirement[];
}

export interface License {
  category?: string;
  status?: string;
  number?: string;
  type?: string;
  issuer?: string;
  expirationDate?: string;
  code?: string;
  state?: string;
  firstName?: string;
  lastName?: string;
  hasBoardAction?: boolean;
  issueDate?: string;
  origin?: string;
  primarySourceCheckedDate?: string;
  primarySourceLastVerifiedDate?: string;
  screenshotId?: string;
  source?: string;
  boardActions?: string[];
  additionalInfo?: {
    deaSchedules?: string;
    licenseState?: string;
    mocStatus?: string;
    renewalDate?: string;
    durationType?: string;
    reverificationDate?: string;
  };
}

export interface NPIValidation {
  npiDetails: {
    providerName: string;
    npi: string;
    updateDate: string;
    providerType?: string;
    Code?: string;
    licenses?: Array<License>;
    entityType?: string;
    enumerationDate?: string;
    gender?: string;
  };
  providerName?: string;
  npi?: string;
  updateDate?: string;
  providerType?: string;
  licenses?: Array<License>;
  entityType?: string;
  enumerationDate?: string;
  rawApiResponse?: any;
}

export interface ProcessedEligibility {
  isEligible: boolean;
  requirements: Requirement[];
  rawValidation: NPIValidation;
  providerType?: string;
  validationMessages?: string[];
}